synapse/handlers/admin.py

```diff
@@ -29,10 +32,106 @@
 class AdminHandler(BaseHandler):
     def __init__(self, hs):
         super(AdminHandler, self).__init__(hs)
-
         self.storage = hs.get_storage()
         self.state_store = self.storage.state
 
+    async def validate_admin_token(
+        self, servlet, request, raise_if_missing: bool = True
+    ) -> bool:
+        """
+        Validate that there is an admin token on the request, and that it can
+        access this servlet.
+        """
+        # This servlet can't be validated by an admin token. Error out.
+        if servlet.PERMISSION_CODE is None:
+            raise SynapseError(403, "Forbidden", errcode=Codes.FORBIDDEN)
+
+        auth_headers = request.requestHeaders.getRawHeaders(b"Authorization")
+
+        if not auth_headers:
+            # No Authorization header, so can't possibly be using an admin token.
+            if raise_if_missing:
+                raise MissingClientTokenError("Missing Authorization header.")
+            else:
+                return False
+
+        if len(auth_headers) > 1:
+            raise MissingClientTokenError("Too many Authorization headers.")
+
+        # Pull out the Bearer token from the Authorization header.
+        parts = auth_headers[0].split(b" ")
+        if parts[0] == b"Bearer" and len(parts) == 2:
+            token = parts[1].decode("ascii")
+        else:
+            raise MissingClientTokenError("Invalid Authorization header.")
+
+        token_rules = await self.store.get_permissions_for_token(token)
+
+        if not raise_if_missing and token_rules.token_state is TokenState.NON_EXISTANT:
+            # The token doesn't exist, but don't raise. The caller is likely to
+            # try to use some other authentication mechanism (e.g. a user
+            # token).
+            return False
+
+        action = request.method.decode("ascii")
+
+        if token_rules.permissions[servlet.PERMISSION_CODE][action] is True:
```

Comment:
```suggestion
        if token_rules.permissions[servlet.PERMISSION_CODE][action]:
```

unless I'm missing something subtle, in which case comment pls.